Detectors/EMCAL/reconstruction/include/EMCALReconstruction/CaloRawFitter.h

```diff
@@ -0,0 +1,156 @@
+
+#include <iosfwd>
+#include <array>
+#include <Rtypes.h>
+#include "EMCALReconstruction/CaloFitResults.h"
+#include "DataFormatsEMCAL/Constants.h"
+#include "EMCALReconstruction/AltroDecoder.h"
+
+namespace o2
+{
+
+namespace emcal
+{
+
+class CaloRawFitter
+{
+
+ public:
+  /// \brief Constructor
+  CaloRawFitter(const char* name, const char* nameshort);
+
+  /// \brief Destructor
+  ~CaloRawFitter() = default;
+
+  /// \brief Method to do the selection of what should possibly be fitted.
+  /// \return Size of the sub-selected sample,
+  /// \return index of the bunch with maximum signal,
+  /// \return maximum signal,
+  /// \return maximum aplitude,
+  /// \return index of max aplitude in array,
+  /// \return pedestal,
+  /// \return first time bin,
+  /// \return last time bin,
+  /// \return ampitude cut
+  std::tuple<int, int, float, short, short, float, int, int> preFitEvaluateSamples(const std::vector<AltroDecoder::Bunch>& bunchvector,
+                                                                                   std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2, int acut);
+
+  /// \brief The require time range if the maximum ADC value is between min and max (timebin)
+  void setTimeConstraint(int min, int max);
+
+  void setIsZeroSuppressed(bool iszs = true) { mIsZerosupressed = iszs; }
+  void setAmpCut(float cut) { mAmpCut = cut; }
+  void setNsamplePed(int i) { mNsamplePed = i; }
+  void setL1Phase(double phase) { mL1Phase = phase; }
+
+  bool getIsZeroSuppressed() const { return mIsZerosupressed; }
+  float getAmpCut() const { return mAmpCut; }
+  int getNsamplePed() const { return mNsamplePed; }
+
+  // access to array info
+  double getReversed(const int i) const { return mReversed[i]; }
+  const char* getAlgoName() const { return mName.c_str(); }
+  const char* getAlgoAbbr() const { return mNameShort.c_str(); }
+  FitAlgorithm getAlgo() const { return mAlgo; }
+
+  /// \brief Get the maximum of a bunch array
+  /// \return Maximum amplitute
+  short maxAmp(const AltroDecoder::Bunch* bunch, int* maxindex) const;
+
+  /// \brief Get maximum of array
+  /// \return Maximum amplitute
+  unsigned short maxAmp(const unsigned short* data, const int length) const;
+
+  /// \brief A bunch is considered invalid if the maximum is in the first or last time-bin.
+  bool checkBunchEdgesForMax(const AltroDecoder::Bunch* const bunch) const;
+
+  /// \brief Check if the index of the max ADC vaue is consistent with trigger.
+  bool isInTimeRange(const int maxindex, const int maxtime, const int mintime) const;
+
+  /// \brief Time sample comes in reversed order, revers them back Subtract the baseline based on content of altrocfg1 and altrocfg2.
+  /// \return Pedestal
+  float reverseAndSubtractPed(const AltroDecoder::Bunch* bunch,
+                              std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2,
+                              double* outarray) const;
```

Comment:
Output parameters (outarray) should not be handled as function parameters. If needed as output better use std::tuple. Furthermore ownership is not clear from the argument type raw pointer.